File: Universe_Backend/Data/Models/Story.cs
using Universe_Backend.Data.DTOs;

namespace Universe_Backend.Data.Models;

public class Story
{
    public int Id { get; set; }
    public required string? Content { get; set; }
    public required string? Image { get; set; }
    public required string? Video { get; set; }
    public required string? Audio { get; set; }
    public DateTime PublishDate { get; set; } = DateTime.Now;
    public required string AuthorId { get; set; }
    public int? SharedPostId { get; set; }
    public int? SharedStoryId { get; set; }
    public ICollection<Widget>? Widgets { get; set; }

    public Post? SharedPost { get; set; }
    public Story? SharedStory { get; set; }
    public virtual User? Author { get; set; }
    public virtual ICollection<User> Mentions { get; set; } = [];
    public virtual ICollection<Tag> Tags { get; set; } = [];
    public virtual ICollection<StoryReaction> Reactions { get; set; } = [];

    public bool IsActive()
    {
        return DateTime.Now.Subtract(PublishDate).TotalHours < 24;
    }

    public StoryDTO ToDTO()
    {
        return new StoryDTO
        {
            Id = Id,
            Content = Content,
            Image = Image,
            Video = Video,
            Audio = Audio,
            PublishDate = PublishDate,
            AuthorId = AuthorId,
            SharedPostId = SharedPostId,
            SharedStoryId = SharedStoryId,
            Widgets = Widgets
        };
    }

    public Story UpdateFromDTO(StoryDTO story)
    {
        Content = story.Content;
        Image = story.Image;
        Video = story.Video;
        Audio = story.Audio;
        PublishDate = story.PublishDate;
        AuthorId = story.AuthorId;
        SharedPostId = story.SharedPostId;
        SharedStoryId = story.SharedStoryId;
        Widgets = story.Widgets;
        return this;
    }
}

File: universe/lib/models/story.dart
import 'package:universe/models/widget.dart';

class Story {
  int id;
  String content;
  String? image;
  String? video;
  String? audio;
  DateTime publishDate;
  int sharedPostId;
  int sharedStoryId;
  List<Widget> widgets;

  Story({
    required this.id,
    required this.content,
    required this.image,
    required this.video,
    required this.audio,
    required this.publishDate,
    required this.sharedPostId,
    required this.sharedStoryId,
    required this.widgets,
  });

  factory Story.fromJson(Map<String, dynamic> json) {
    return Story(
      id: json['id'],
      content: json['content'],
      image: json['image'],
      video: json['video'],
      audio: json['audio'],
      publishDate: DateTime.parse(json['publishDate']),
      sharedPostId: json['sharedPostId'],
      sharedStoryId: json['sharedStoryId'],
      widgets: json['widgets'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'content': content,
      'image': image,
      'video': video,
      'audio': audio,
      'publishDate': publishDate.toIso8601String(),
      'sharedPostId': sharedPostId,
      'sharedStoryId': sharedStoryId,
      'widgets': widgets,
    };
  }
}
